common/test/index_test.cpp

```diff
@@ -3,6 +3,152 @@
 #include <gtest/gtest.h>
 
 #include "scipp/common/index.h"
+#include "scipp/common/index_composition.h"
 
 TEST(IndexTest, size) { ASSERT_EQ(sizeof(scipp::index), 8); }
 TEST(IndexTest, sign) { ASSERT_EQ(scipp::index{-1}, int64_t(-1)); }
+
+TEST(IndexCompositionTest, flat_index_from_strides_0d) {
+  const std::array<scipp::index, 0> strides{};
+  const std::array<scipp::index, 0> indices{};
+  EXPECT_EQ(scipp::flat_index_from_strides(strides.begin(), strides.end(),
+                                           indices.begin()),
+            0);
+}
+
+TEST(IndexCompositionTest, flat_index_from_strides_1d) {
+  for (scipp::index stride = 0; stride < 6; ++stride) {
+    const std::array<scipp::index, 1> strides{stride};
+    for (scipp::index index = 0; index < stride; ++index) {
+      const std::array<scipp::index, 1> indices{index};
+      EXPECT_EQ(scipp::flat_index_from_strides(strides.begin(), strides.end(),
+                                               indices.begin()),
+                index * stride);
+    }
+  }
+}
+
+TEST(IndexCompositionTest, flat_index_from_strides_2d) {
+  for (scipp::index stride0 = 0; stride0 < 6; ++stride0) {
+    for (scipp::index stride1 = 0; stride1 < 6; ++stride1) {
+      const std::array<scipp::index, 2> strides{stride0, stride1};
+      for (scipp::index index0 = 0; index0 < stride0; ++index0) {
+        for (scipp::index index1 = 0; index1 < stride1; ++index1) {
+          const std::array<scipp::index, 2> indices{index0, index1};
+          EXPECT_EQ(scipp::flat_index_from_strides(
+                        strides.begin(), strides.end(), indices.begin()),
+                    index0 * stride0 + index1 * stride1);
+        }
+      }
+    }
+  }
+}
+
+TEST(IndexCompositionTest, flat_index_from_strides_3d) {
+  for (scipp::index stride0 = 0; stride0 < 6; ++stride0) {
+    for (scipp::index stride1 = 0; stride1 < 6; ++stride1) {
+      for (scipp::index stride2 = 0; stride2 < 6; ++stride2) {
+        const std::array<scipp::index, 3> strides{stride0, stride1, stride2};
+        for (scipp::index index0 = 0; index0 < stride0; ++index0) {
+          for (scipp::index index1 = 0; index1 < stride1; ++index1) {
+            for (scipp::index index2 = 0; index2 < stride2; ++index2) {
+              const std::array<scipp::index, 3> indices{index0, index1, index2};
+              EXPECT_EQ(scipp::flat_index_from_strides(
+                            strides.begin(), strides.end(), indices.begin()),
+                        index0 * stride0 + index1 * stride1 + index2 * stride2);
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+TEST(IndexCompositionTest, extract_indices_0d) {
+  const std::array<scipp::index, 3> shape{-1, -2, -3};
```

Comment:
What is the meaning of negative values in `shape`?

Reply:
To catch it `extract_indices` overwrites the correct values and nothing else because it should never write negative numbers.

Reply:
I don't understand you reply, can you reformulate?

Reply:
I mixed things up. The indices start with negative values to ensure that `extract_indices` overwrites them. Initialising with 0 would not catch the case where 0 is the correct answer but `extract_indices` fails to write it out.
For `shape`, this is just a way or marking unused elements. And maybe trigger an error if the function uses those elements, though that should happen regardless of what value is in there.

Reply:
👍